sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -1239,6 +1260,303 @@ ProgramManager::getSYCLDeviceImages(const context &Ctx,
+      std::make_shared<detail::device_image_impl>(Context, Devs,
+                                                  bundle_state::object);
+
+  ExecutableImpl->get_program_ref() = LinkedProg;
+
+  std::vector<kernel_id> KernelIDs;
+  for (const device_image_plain &DeviceImage : DeviceImages) {
+    // Duplicates are not expected here, otherwise piProgramLink should fail
+    KernelIDs.insert(KernelIDs.end(),
+                     getSyclObjImpl(DeviceImage)->get_kernel_ids().begin(),
+                     getSyclObjImpl(DeviceImage)->get_kernel_ids().end());
+  }
+
+  ExecutableImpl->get_kernel_ids_ref() = KernelIDs;
+
+  return {createSyclObjFromImpl<device_image_plain>(ExecutableImpl)};
+}
+
+device_image_plain ProgramManager::build(const device_image_plain &DeviceImage,
+                                         const std::vector<device> &Devs,
+                                         const property_list &PropList) {
+  (void)PropList;
+
+  const std::shared_ptr<device_image_impl> &InputImpl =
+      getSyclObjImpl(DeviceImage);
+
+  const context Context = InputImpl->get_context();
+
+  // TODO: assert state
+  // TODO: Devs should be compatible with device image
+  // TODO: DeviceImage state should be input
+
+  DeviceImageImplPtr ExecImpl = std::make_shared<detail::device_image_impl>(
+      InputImpl->get_bin_image_ref(), Context, Devs, bundle_state::executable);
+
+  //void *M = (void*)(-1);
+  // TODO: Make sure that KSIds will be different for the case when the same
+  // kernel built with different options is present in the fat binary.
+  //KernelSetId KSId = getKernelSetId(M, KernelName);
+  //
+
+  const ContextImplPtr ContextImpl = getSyclObjImpl(Context);
+
+  using PiProgramT = KernelProgramCache::PiProgramT;
+  using ProgramCacheT = KernelProgramCache::ProgramCacheT;
+
+  KernelProgramCache &Cache = ContextImpl->getKernelProgramCache();
+
+  auto AcquireF = [](KernelProgramCache &Cache) {
+    return Cache.acquireCachedPrograms();
+  };
+  auto GetF = [](const Locked<ProgramCacheT> &LockedCache) -> ProgramCacheT & {
+    return LockedCache.get();
+  };
+
+  std::string CompileOpts;
+  std::string LinkOpts;
+  // Build options are overridden if environment variables are present.
+  // Environment variables are not changed during program lifecycle so it
+  // is reasonable to use static here to read them only once.
+  static const char *CompileOptsEnv =
+      SYCLConfig<SYCL_PROGRAM_COMPILE_OPTIONS>::get();
+  if (CompileOptsEnv)
+    CompileOpts = CompileOptsEnv;
+
+  static const char *LinkOptsEnv = SYCLConfig<SYCL_PROGRAM_LINK_OPTIONS>::get();
+  if (LinkOptsEnv) {
+    LinkOpts = LinkOptsEnv;
+  }
+
+  RTDeviceBinaryImage *ImgPtr = InputImpl->get_bin_image_ref();
+  const RTDeviceBinaryImage &Img = *ImgPtr;
+
+  auto BuildF = [this, &Context, Img, &Devs, &CompileOpts, &LinkOpts,
+                 &ExecImpl] {
+    // Update only if compile options are not overwritten by environment
+    // variable
+    if (!CompileOptsEnv) {
```

Comment:
These code partially overlaps with what program_manager already does in other functions. But those functions do not work with multiple devices.
Changing whole program_manager in scope of this patch is too intrusive, so going to refactor it separately.

Reply:
Does it make sense to add a TODO comment?

Reply:
Added